MoreLinq/ToArrayByIndex.cs

```diff
@@ -0,0 +1,241 @@
+        /// <returns>
+        /// An array of size <paramref name="length"/> that contains the
+        /// elements from the input sequence.
+        /// </returns>
+        /// <remarks>
+        /// This method forces immediate query evaluation. It should not be
+        /// used on infinite sequences. If more than one element maps to the
+        /// same index then the latter element overwrites the former in the
+        /// resulting array.
+        /// </remarks>
+
+        public static T[] ToArrayByIndex<T>(this IEnumerable<T> source, int length,
+            Func<T, int> indexSelector)
+        {
+            return source.ToArrayByIndex(length, indexSelector, (e, _) => e);
+        }
+
+        /// <summary>
+        /// Creates an array of user-specified length from an
+        /// <see cref="IEnumerable{T}"/> where a function is used to determine
+        /// the index at which an element will be placed in the array. The
+        /// elements are projected into the array via an additional function.
+        /// </summary>
+        /// <param name="source">The source sequence for the array.</param>
+        /// <param name="length">The length of the resulting array.</param>
+        /// <param name="indexSelector">
+        /// A function that maps an element to its index.</param>
+        /// <param name="resultSelector">
+        /// A function to project a source element into an element of the
+        /// resulting array.</param>
+        /// <typeparam name="T">
+        /// The type of the element in<paramref name="source"/>.</typeparam>
+        /// <typeparam name="TResult">
+        /// The type of the element in the resulting arrary.</typeparam>
+        /// <returns>
+        /// An array of size <paramref name="length"/> that contains the
+        /// projected elements from the input sequence.
+        /// </returns>
+        /// <remarks>
+        /// This method forces immediate query evaluation. It should not be
+        /// used on infinite sequences. If more than one element maps to the
+        /// same index then the latter element overwrites the former in the
+        /// resulting array.
+        /// </remarks>
+
+        public static TResult[] ToArrayByIndex<T, TResult>(this IEnumerable<T> source, int length,
+            Func<T, int> indexSelector, Func<T, TResult> resultSelector)
+        {
+            return source.ToArrayByIndex(length, indexSelector, (e, _) => resultSelector(e));
+        }
+
+        /// <summary>
+        /// Creates an array of user-specified length from an
+        /// <see cref="IEnumerable{T}"/> where a function is used to determine
+        /// the index at which an element will be placed in the array. The
+        /// elements are projected into the array via an additional function.
+        /// </summary>
+        /// <param name="source">The source sequence for the array.</param>
+        /// <param name="length">The length of the resulting array.</param>
```

Comment:
Hmm…I wonder why [`Enumerable.Take`][take] even allows count to be negative.

> If count is less than or equal to zero, source is not enumerated and an empty `IEnumerable<T>` is returned.

Wonder if same spirit should be maintained here although I am tempted to validate the argument.

[take]: https://msdn.microsoft.com/en-us/library/bb503062.aspx